pytorch_lightning/callbacks/timer.py

```diff
@@ -0,0 +1,117 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Timer
+^^^^^
+"""
+import logging
+from datetime import datetime, timedelta
+from typing import Any, Dict, Union
+
+from pytorch_lightning.callbacks.base import Callback
+from pytorch_lightning.utilities.distributed import rank_zero_info
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+log = logging.getLogger(__name__)
+
+
+class Timer(Callback):
+    """
+    The Timer callback tracks the time spent in the training loop and interrupts the Trainer
+    if the given time limit is reached.
+
+    Args:
+        duration: A string in the format HH:MM:SS (hours, minutes seconds), or a :class:`datetime.timedelta`.
+        interval: Determines if the interruption happens on epoch level or mid-epoch.
+            Can be either `epoch` or `step`.
+        verbose: Set this to ``False`` to suppress logging messages.
+
+    Raises:
+        MisconfigurationException:
+            If ``interval`` is not one of the supported choices.
+
+    Example::
+        from pytorch_lightning import Trainer
+        from pytorch_lightning.callbacks import Timer
+
+        # stop training after 12 hours
+        timer = Timer(duration="12:00:00")
+
+        # or provide a datetime.timedelta
+        from datetime import timedelta
+        timer = Timer(duration=timedelta(weeks=1))
+
+        # force training to stop after given time limit
+        trainer = Trainer(callbacks=[timer])
+    """
+
+    INTERVAL_CHOICES = ("epoch", "step")
+
+    def __init__(self, duration: Union[str, timedelta], interval: str = "step", verbose: bool = True):
+        super().__init__()
+        if isinstance(duration, str):
+            hms = datetime.strptime(duration.strip(), "%H:%M:%S")
+            duration = timedelta(hours=hms.hour, minutes=hms.minute, seconds=hms.second)
+        if interval not in self.INTERVAL_CHOICES:
+            raise MisconfigurationException(
+                f"Unsupported parameter value `Timer(interval={interval})`. Possible choices are:"
+                f" {', '.join(self.INTERVAL_CHOICES)}"
+            )
+        self._duration = duration
+        self._interval = interval
+        self._verbose = verbose
+        self._start_time = None
+        self._offset = timedelta()
+
+    @property
+    def start_time(self):
+        return self._start_time
+
+    @property
+    def time_elapsed(self) -> timedelta:
+        if self._start_time is None:
+            return self._offset
+        return datetime.now() - self._start_time + self._offset
+
+    @property
+    def time_remaining(self) -> timedelta:
+        return self._duration - self.time_elapsed
+
+    def on_train_start(self, trainer, *args, **kwargs) -> None:
```

Comment:
I would create `_start_time` on `__init__` instead.

Reply:
The user might want to run `.test()` or `.validate()` first before running the training.
Do we want to include this or only training?